chain/store/name.go

```diff
@@ -1,63 +1,150 @@
 package store
 
 import (
-	"strings"
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"sort"
+
+	"github.com/nknorg/nkn/common/serialization"
 )
 
-func getRegistrantId(registrant []byte) string {
-	return string(registrant)
+type nameInfo struct {
+	registrant []byte
+	expiresAt  uint32
 }
 
-func getNameId(name string) string {
-	return strings.ToLower(name)
+type namesCleanup map[string]struct{}
+
+func newNameInfo(registrant []byte, expiresAt uint32) *nameInfo {
+	ni := new(nameInfo)
+	ni.registrant = registrant
+	ni.expiresAt = expiresAt
+	return ni
 }
 
-func (sdb *StateDB) updateName(registrant []byte, name string) error {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) Serialize(w io.Writer) error {
+	if err := serialization.WriteVarBytes(w, ni.registrant); err != nil {
+		return err
+	}
+	if err := serialization.WriteUint32(w, ni.expiresAt); err != nil {
+		return err
+	}
+	return nil
+}
 
-	err := sdb.trie.TryUpdate(append(NamePrefix, nameId...), registrant)
+func (ni *nameInfo) Deserialize(r io.Reader) error {
+	var err error
+	ni.registrant, err = serialization.ReadVarBytes(r)
+	if err != nil {
+		return err
+	}
+	ni.expiresAt, err = serialization.ReadUint32(r)
 	if err != nil {
 		return err
 	}
+	return nil
+}
 
-	return sdb.trie.TryUpdate(append(NameRegistrantPrefix, registrantId...), []byte(name))
+func (ni *nameInfo) Bytes() []byte {
+	buff := bytes.NewBuffer(nil)
+	ni.Deserialize(buff)
+	return buff.Bytes()
 }
 
-func (sdb *StateDB) setName(registrant []byte, name string) {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) Empty() bool {
+	return len(ni.registrant) == 0 && ni.expiresAt == 0
+}
 
-	sdb.names.Store(registrantId, name)
-	sdb.nameRegistrants.Store(nameId, registrant)
+func getRegistrantId(registrant []byte) string {
+	return string(registrant)
 }
 
-func (sdb *StateDB) deleteName(registrantId string) error {
-	name, err := sdb.trie.TryGet(append(NameRegistrantPrefix, registrantId...))
+func getNameId(name string, registrant []byte, expiresAt uint32) []byte {
+	nameId := bytes.NewBuffer(nil)
+	nameId.Write([]byte(name))
+
+	ni := newNameInfo(registrant, expiresAt)
+	registrantHash := sha256.Sum256(ni.Bytes())
```

Comment:
Don't use obj.`Bytes()` as input of Hash(), used obj.`Deserialize()` for consistency between platforms.